beacon_node/lighthouse_network/src/peer_manager/connectivity.rs

```diff
@@ -0,0 +1,188 @@
+            //    in the next heartbeat down to our target).
+            // 2. If the peer is one our validators require for a specific subnet, then it is
+            //    considered a priority. We have pre-allocated some extra priority slots for these
+            //    peers as specified by PRIORITY_PEER_EXCESS. Therefore we dial these peers, even
+            //    if we are already at our max_peer limit.
+            if !self.peers_to_dial.contains(enr)
+                && ((min_ttl.is_some()
+                    && connected_or_dialing + to_dial_peers < self.max_priority_peers())
+                    || connected_or_dialing + to_dial_peers < self.max_peers())
+            {
+                // This should be updated with the peer dialing. In fact created once the peer is
+                // dialed
+                let peer_id = enr.peer_id();
+                if let Some(min_ttl) = min_ttl {
+                    self.network_globals_provider
+                        .update_min_ttl(&peer_id, *min_ttl);
+                }
+                if self.dial_peer(enr.clone()) {
+                    //debug!(self.log, "Added discovered ENR peer to dial queue"; "peer_id" => %peer_id);
+                    to_dial_peers += 1;
+                }
+            }
+        }
+
+        // The heartbeat will attempt new discovery queries every N seconds if the node needs more
+        // peers. As an optimization, this function can recursively trigger new discovery queries
+        // immediatelly if we don't fulfill our peers needs after completing a query. This
+        // recursiveness results in an infinite loop in networks where there not enough peers to
+        // reach out target. To prevent the infinite loop, if a query returns no useful peers, we
+        // will cancel the recursiveness and wait for the heartbeat to trigger another query latter.
+        // The heartbeat will attempt new discovery queries every N seconds if the node needs more
+        // peers. As an optimization, this function can recursively trigger new discovery queries
+        // immediatelly if we don't fulfill our peers needs after completing a query. This
+        // recursiveness results in an infinite loop in networks where there not enough peers to
+        // reach out target. To prevent the infinite loop, if a query returns no useful peers, we
+        // will cancel the recursiveness and wait for the heartbeat to trigger another query latter.
+        if results_count > 0 && to_dial_peers == 0 {
+            //debug!(self.log, "Skipping recursive discovery query after finding no useful results"; "results" => results_count);
+            crate::metrics::inc_counter(&crate::metrics::DISCOVERY_NO_USEFUL_ENRS);
+            0
+        } else {
+            // Queue another discovery if we need to
+            self.maintain_peer_count(to_dial_peers)
+        }
+    }
+
+    /// This function checks the status of our current peers and optionally requests a discovery
```

Comment:
These comments are now longer correct. This function returns a usize and doesn't request any discoveries. Maybe we need to change the name of the function to better reflect this change of logic.